jobsdb/jobsdb.go

```diff
@@ -1350,94 +1349,23 @@ func (jd *HandleT) doComputeNewIdxForAppend(dList []dataSetT) string {
-			return calculatedVals, err
-		}
-		// Just increment the last value of the index as a possible candidate
-		calculatedVals[len(calculatedVals)-1] = fmt.Sprintf("%d", lastVal+1)
-		var equals bool
-		if len(calculatedVals) == len(after) {
-			equals = true
-			for k := 0; k < len(calculatedVals); k++ {
-				if calculatedVals[k] == after[k] {
-					continue
-				}
-				equals = false
-			}
-		}
-		if !equals {
-			comparison, err := dsComparitor(calculatedVals, after)
-			if err != nil {
-				return calculatedVals, err
-			}
-			if !comparison {
-				return calculatedVals, fmt.Errorf("computed index is invalid. before: %v, after: %v, calculatedVals: %v", before, after, calculatedVals)
-			}
-		}
-		// Only when the index starts with 0, we allow three levels and when we are using the legacy migration
-		// In all other cases, we allow only two levels
-		var comparatorBool bool
-		if skipZeroAssertionForMultitenant {
-			comparatorBool = len(calculatedVals) == 2
-		} else {
-			comparatorBool = (before[0] == "0" && len(calculatedVals) == 3) ||
-				(before[0] != "0" && len(calculatedVals) == 2)
-		}
-
-		if comparatorBool {
-			if equals {
-				return calculatedVals, fmt.Errorf("calculatedVals and after are same. computed index is invalid. before: %v, after: %v, calculatedVals: %v", before, after, calculatedVals)
-			} else {
-				return calculatedVals, nil
-			}
-		}
-
-		before = append(before, "0")
-	}
-}
-
 func computeInsertIdx(beforeIndex, afterIndex string) (string, error) {
-	comparison, err := dsComparitor(strings.Split(beforeIndex, "_"), strings.Split(afterIndex, "_"))
+	before, err := dsindex.Parse(beforeIndex)
 	if err != nil {
-		return "", fmt.Errorf("Error while comparing beforeIndex: %s and afterIndex: %s with error : %w", beforeIndex, afterIndex, err)
+		return "", fmt.Errorf("could not parse before index: %w", err)
 	}
-	if !comparison {
-		return "", fmt.Errorf("Not a valid insert request between %s and %s", beforeIndex, afterIndex)
-	}
-
-	// No dataset should have 0 as the index.
-	// 0_1, 0_2 are allowed.
-	if beforeIndex == "0" && !skipZeroAssertionForMultitenant {
-		return "", fmt.Errorf("Unsupported beforeIndex: %s", beforeIndex)
+	after, err := dsindex.Parse(afterIndex)
+	if err != nil {
+		return "", fmt.Errorf("could not parse after index: %w", err)
 	}
-
-	beforeVals := strings.Split(beforeIndex, "_")
-	afterVals := strings.Split(afterIndex, "_")
-	calculatedInsertVals, err := computeInsertVals(beforeVals, afterVals)
+	result, err := before.Bump(after)
 	if err != nil {
-		return "", fmt.Errorf("Failed to calculate InserVals with error: %w", err)
+		return "", fmt.Errorf("could not compute insert index: %w", err)
 	}
-	calculatedIdx := strings.Join(calculatedInsertVals, "_")
-	if len(calculatedInsertVals) > 3 {
-		return "", fmt.Errorf("We don't expect a ds to be computed to Level3. We got %s while trying to insert between %s and %s", calculatedIdx, beforeIndex, afterIndex)
+	if result.Length() > 2 {
```

Comment:
Why should we explicitly forbid a level 3 index?
In practice we know that this will not happen, but what would be the problem even if it happened?
As long as we can guarantee that the calculated index is still smaller than the `afterIndex` and greater than the `beforeIndex` we should accept the generated index.

Reply:
Yes… it would read the smallest DS index and create tables accordingly

Reply:
then even a 10 lvl index shouldn't complicate things, would it?